bt-core/src/main/java/bt/torrent/messaging/Assignment.java

```diff
@@ -72,6 +72,13 @@ ConnectionKey getConnectionKey() {
     private void claimPiecesIfNeeded() {
         if (pieces.size() < MAX_SIMULTANEOUSLY_ASSIGNED_PIECES) {
             Bitfield peerBitfield = pieceStatistics.getPeerBitfield(connectionKey).get();
+            if(!assignments.isEndgame()){
```

Comment:
So this is same as below, but without shuffling?

Reply:
it's the same as the code below, shuffling applied only for 'end-game' as I understand, the point of this change is to read only up to 3 values from the stream, not the whole stream and place it into array